src/compiler/parser.rs

```diff
@@ -629,18 +882,130 @@ impl ParserType {
         }
     }
 
     fn parse_statement(&mut self) -> Statement {
         match self.token() {
-            SyntaxKind::SemicolonToken => self.parse_empty_statement(),
-            _ => self.parse_expression_or_labeled_statement(),
+            SyntaxKind::SemicolonToken => return self.parse_empty_statement(),
+            SyntaxKind::ConstKeyword => {
+                if self.is_start_of_declaration() {
+                    return self.parse_declaration();
+                }
+            }
+            _ => (),
         }
+        return self.parse_expression_or_labeled_statement();
+    }
+
+    fn parse_declaration(&mut self) -> Statement {
+        let pos = self.get_node_pos();
+        if false {
+            unimplemented!()
+        } else {
+            self.parse_declaration_worker(pos)
+        }
+    }
+
+    fn parse_declaration_worker(&mut self, pos: usize) -> Statement {
+        match self.token() {
+            SyntaxKind::ConstKeyword => self.parse_variable_statement(pos),
+            _ => unimplemented!(),
+        }
+    }
+
+    fn is_binding_identifier_or_private_identifier_or_pattern(&self) -> bool {
+        self.is_binding_identifier()
+    }
+
+    fn parse_identifier_or_pattern(&mut self) -> Rc<Node> {
+        Rc::new(self.parse_binding_identifier().into())
+    }
+
+    fn parse_variable_declaration_no_exclamation(&mut self) -> VariableDeclaration {
+        self.parse_variable_declaration(false)
+    }
+
+    fn parse_variable_declaration_allow_exclamation(&mut self) -> VariableDeclaration {
+        self.parse_variable_declaration(true)
+    }
+
+    fn parse_variable_declaration(&mut self, allow_exclamation: bool) -> VariableDeclaration {
+        let pos = self.get_node_pos();
+        let name = self.parse_identifier_or_pattern();
+        let type_ = self.parse_type_annotation();
+        let initializer = if false {
+            None
+        } else {
+            self.parse_initializer()
+        };
+        let node = self.factory.create_variable_declaration(
+            self,
+            Some(name),
+            type_.map(|type_| Rc::new(type_.into())),
+            initializer.map(|initializer| Rc::new(initializer.into())),
+        );
+        self.finish_node(node, pos, None)
+    }
+
+    fn parse_variable_declaration_list(&mut self) -> VariableDeclarationList {
+        let pos = self.get_node_pos();
+
+        let mut flags = NodeFlags::None;
+        match self.token() {
+            SyntaxKind::ConstKeyword => {
+                flags |= NodeFlags::Const;
+            }
+            _ => Debug_.fail(None),
+        }
+
+        self.next_token();
+
+        let declarations: NodeArray;
+        if false {
+            unimplemented!()
+        } else {
+            declarations = self.parse_delimited_list(
+                ParsingContext::VariableDeclarations,
+                ParserType::parse_variable_declaration_allow_exclamation,
+            );
```

Comment:
I'm not sure I realized you could do this (declare a variable but not set it right away). A maybe slightly more idiomatic way to do that would be to `let declarations = if ...`, but maybe that doesn't work with the future code of that if-statement.

Reply:
Ya I wasn't sure if Rust supported delayed initialization like this but then ya this mimics the Typescript version pretty closely